<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />



<title>datamodel</title>

<script src="libs/jquery/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="libs/bootstrap/css/flatly.min.css" rel="stylesheet" />
<script src="libs/bootstrap/js/bootstrap.min.js"></script>
<script src="libs/bootstrap/shim/html5shiv.min.js"></script>
<script src="libs/bootstrap/shim/respond.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="libs/highlight/default.css"
      type="text/css" />
<script src="libs/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img { 
  max-width:100%; 
  height: auto; 
}
</style>
<div class="container-fluid main-container">

<style type="text/css">
  body {
    padding-top: 20px;
  }
  .navbar-brand {
    font-weight: bold;
  }
  footer {
    border-top: 1px solid #CCC;
    margin-top: 60px;
    margin-bottom: 48px;
    opacity: 0.75;
  }
  .content-body>:first-child, .content-body>:first-child>:first-child {
    margin-top: 0 !important;
  }
  p code {
    white-space: inherit;
  }
</style>

<div class="fluid-row">
  <div class="col-sm-12">
    <nav class="navbar navbar-inverse">
      <div class="container-fluid">
        <!-- Brand and toggle get grouped for better mobile display -->
        <div class="navbar-header">
          <a class="navbar-brand" href="./">phylogeo: geographic analysis of microbiome data</a>
        </div>
      </div>
    </nav>
  </div>
</div>

<div class="fluid-row">
  <div class="col-sm-3">
    <div class="list-group">
      <a class="list-group-item" href="./">Introduction</a>
      <a class="list-group-item" href="mapping_functions.html">Mapping Functions</a>
      <a class="list-group-item" href="plotting_functions.html">Plotting Functions</a>
      <a class="list-group-item" href="projections.html">Projections</a>
      <a class="list-group-item" href="datamodel.html">Data Model</a>
      
      <!--<a class="list-group-item" href="raster.html">Raster Images</a>-->
      </div>
    <script>
      // manage active state of toc based on current page
      var href = window.location.pathname;
      href = href.substr(href.lastIndexOf('/') + 1);
      if (href === "" || href === "index.html")
        href = "./";
      $('a.list-group-item[href="' + href + '"]').addClass('active');
    </script>
  </div>
  <div class="col-sm-9 content-body">



<p>phylogeo is based on top of phyloseq and therefore uses the same data model, an excellent overview of which can be found here. This vignette is simply a continuation of that tutorial. We create some random data from phyloseq including the following:</p>
<ul>
<li>an OTU table,</li>
<li>a Taxonomy Table,</li>
<li>a Sample Data dataframe.</li>
<li>Latitude and Longitude information that is added to the Sample Data, and</li>
<li>A phylogenetic tree of the relationship betwen the OTUs contianed in the OTU table.</li>
</ul>
<p>Phyloseq provides tools for dealing with the first three items on this list. By adding location information and phylogenetic information you can take advantage fo the plotting tools in the phylogeo package.</p>
<p>Note: If these data-types don’t make sense to you, it may be worthwhile to look at the phyloseq tutorials to get your bearings. For a grasp of what OTUs are and how they are used in metagenome analysis you could do worse than to start with Robert Edgars’ page on OTUs. Edgar is the author of a number of highly used bioinformatic programs that comarse DNA sequences including Muscle and USeqrch (see here).</p>
<div id="generate-phyloseq-data-and-plot-with-phylogeo" class="section level2">
<h2>Generate phyloseq data and plot with phylogeo</h2>
<p>Load phylogeo and get some sequecne data from the ape package</p>
<p>In this vignette, sequence data from the ape package will be used, although you will probably have your own sequencing data to use. Everything except that sequence data is generated denovo using base R.</p>
</div>
<div id="create-an-otu-table" class="section level2">
<h2>Create an OTU table</h2>
<p>An operational taxonomic unit (OTU) table that a table holding count information. The row values are typicaly OTUs which in macro-biology are typically species, but in microbiology are semi-arbitrary cutoffs of sequence similarity that corresponds roughly to species. For 16S studies, these OTUs are typically 16S DNA amplicon sequences clustered at 97% identity. We can make an arbitrary dataset where out OTUs will be the woodmouse data, and the columns will be our samples.</p>
<pre class="r"><code>otumat = matrix(sample(1:100, 75, replace = TRUE), 
                nrow = length(woodmousenames), 
                ncol = samplenum)
rownames(otumat) &lt;- woodmousenames
colnames(otumat) &lt;- samplenames
otumat</code></pre>
<pre><code>##         Sample1 Sample2 Sample3 Sample4 Sample5 Sample6 Sample7 Sample8
## No305        50      71      91      17       5      50      71      91
## No304        79      37       6      49      48      79      37       6
## No306        55      98      84      37      70      55      98      84
## No0906S       9      85      97      80      14       9      85      97
## No0908S      43      93      27      43      75      43      93      27
## No0909S      98      62       5      53      44      98      62       5
## No0910S      34      55      86      54       2      34      55      86
## No0912S      90      62      71      17      89      90      62      71
## No0913S      42      27      39      57      60      42      27      39
## No1103S      89      36      78      38      99      89      36      78
## No1007S      67      11      64      87      58      67      11      64
## No1114S      82      41      42      34      96      82      41      42
## No1202S      44      41      38      98      33      44      41      38
## No1206S      61      45      10      94      36      61      45      10
## No1208S      91      40      38      92      48      91      40      38
##         Sample9 Sample10 Sample11 Sample12 Sample13 Sample14 Sample15
## No305        17        5       50       71       91       17        5
## No304        49       48       79       37        6       49       48
## No306        37       70       55       98       84       37       70
## No0906S      80       14        9       85       97       80       14
## No0908S      43       75       43       93       27       43       75
## No0909S      53       44       98       62        5       53       44
## No0910S      54        2       34       55       86       54        2
## No0912S      17       89       90       62       71       17       89
## No0913S      57       60       42       27       39       57       60
## No1103S      38       99       89       36       78       38       99
## No1007S      87       58       67       11       64       87       58
## No1114S      34       96       82       41       42       34       96
## No1202S      98       33       44       41       38       98       33
## No1206S      94       36       61       45       10       94       36
## No1208S      92       48       91       40       38       92       48</code></pre>
</div>
<div id="now-we-need-a-pretend-taxonomy-table" class="section level2">
<h2>Now we need a pretend taxonomy table</h2>
<p>The taxonomy table lets you assign taxonomic information for each OTU. The rownames of the taxonomy table must be the same as the OTU table. Each of the columns represent some value, typically corresponding to the classic Kindom, Phylum…. Species heirarchy. However, this table can be used to assign arbitrary data to OTUs as well. Here we will generate some random taxonomy data using classic phylogeny, but with some made-up, filler data.</p>
<pre class="r"><code>taxmat = matrix(sample(letters, 70, replace = TRUE), 
                nrow = nrow(otumat), 
                ncol = 7)
rownames(taxmat) &lt;- rownames(otumat)
colnames(taxmat) &lt;- c(&quot;Domain&quot;, &quot;Phylum&quot;, &quot;Class&quot;, &quot;Order&quot;, 
                      &quot;Family&quot;, &quot;Genus&quot;, &quot;Species&quot;)
taxmat</code></pre>
<pre><code>##         Domain Phylum Class Order Family Genus Species
## No305   &quot;n&quot;    &quot;k&quot;    &quot;m&quot;   &quot;j&quot;   &quot;n&quot;    &quot;d&quot;   &quot;r&quot;    
## No304   &quot;a&quot;    &quot;z&quot;    &quot;v&quot;   &quot;h&quot;   &quot;n&quot;    &quot;n&quot;   &quot;h&quot;    
## No306   &quot;z&quot;    &quot;n&quot;    &quot;f&quot;   &quot;o&quot;   &quot;s&quot;    &quot;n&quot;   &quot;h&quot;    
## No0906S &quot;j&quot;    &quot;d&quot;    &quot;h&quot;   &quot;p&quot;   &quot;w&quot;    &quot;e&quot;   &quot;k&quot;    
## No0908S &quot;r&quot;    &quot;j&quot;    &quot;s&quot;   &quot;a&quot;   &quot;i&quot;    &quot;g&quot;   &quot;x&quot;    
## No0909S &quot;d&quot;    &quot;r&quot;    &quot;c&quot;   &quot;z&quot;   &quot;g&quot;    &quot;k&quot;   &quot;j&quot;    
## No0910S &quot;n&quot;    &quot;h&quot;    &quot;i&quot;   &quot;r&quot;   &quot;q&quot;    &quot;t&quot;   &quot;w&quot;    
## No0912S &quot;n&quot;    &quot;h&quot;    &quot;j&quot;   &quot;z&quot;   &quot;d&quot;    &quot;r&quot;   &quot;m&quot;    
## No0913S &quot;e&quot;    &quot;k&quot;    &quot;y&quot;   &quot;x&quot;   &quot;h&quot;    &quot;g&quot;   &quot;c&quot;    
## No1103S &quot;g&quot;    &quot;x&quot;    &quot;i&quot;   &quot;u&quot;   &quot;i&quot;    &quot;e&quot;   &quot;z&quot;    
## No1007S &quot;k&quot;    &quot;j&quot;    &quot;h&quot;   &quot;u&quot;   &quot;n&quot;    &quot;k&quot;   &quot;m&quot;    
## No1114S &quot;t&quot;    &quot;w&quot;    &quot;e&quot;   &quot;k&quot;   &quot;a&quot;    &quot;z&quot;   &quot;v&quot;    
## No1202S &quot;r&quot;    &quot;m&quot;    &quot;n&quot;   &quot;h&quot;   &quot;z&quot;    &quot;n&quot;   &quot;f&quot;    
## No1206S &quot;g&quot;    &quot;c&quot;    &quot;a&quot;   &quot;f&quot;   &quot;j&quot;    &quot;d&quot;   &quot;h&quot;    
## No1208S &quot;e&quot;    &quot;z&quot;    &quot;g&quot;   &quot;n&quot;   &quot;r&quot;    &quot;j&quot;   &quot;s&quot;</code></pre>
</div>
<div id="and-some-fake-sample-data" class="section level2">
<h2>And some fake Sample Data</h2>
<p>Sample data is information about your samples. The sample data rownames must corresond to the values in the OTU table. We will make some smaple data with some arbitrary columns. These columns are often used to store informaiton about sample processing, or about the samples themselves.</p>
<pre class="r"><code>sampledata &lt;- matrix(sample(letters,samplenum*7, replace=TRUE), 
                     nrow = samplenum, 
                     ncol = 7)
rownames(sampledata) &lt;- samplenames
# in your real data these columns might hold soil data, or processing information
colnames(sampledata) &lt;- paste0(&quot;Category&quot;, 1:ncol(sampledata))
# generate some random lat and lon data and add it to the smaple data.
lat = sample((-90:90), samplenum)
lon = sample((-180:180), samplenum)
sampledata &lt;- cbind(sampledata, lat, lon)
sampledata &lt;- as.data.frame(sampledata)
sampledata</code></pre>
<pre><code>##          Category1 Category2 Category3 Category4 Category5 Category6
## Sample1          n         k         g         z         p         l
## Sample2          w         b         v         d         p         m
## Sample3          w         r         o         j         l         r
## Sample4          y         b         h         d         p         s
## Sample5          z         l         i         h         n         u
## Sample6          g         k         m         n         j         o
## Sample7          b         r         c         i         k         t
## Sample8          p         u         d         s         o         d
## Sample9          l         a         h         c         v         c
## Sample10         y         j         g         y         x         v
## Sample11         l         u         d         l         v         e
## Sample12         w         a         j         b         p         a
## Sample13         k         k         y         h         m         l
## Sample14         k         m         r         a         o         h
## Sample15         n         z         m         c         f         j
##          Category7 lat lon
## Sample1          s -32 172
## Sample2          a  38  34
## Sample3          y -31 -57
## Sample4          v  23 -23
## Sample5          s  86 137
## Sample6          i -46  -6
## Sample7          t -43 -99
## Sample8          n -85  -9
## Sample9          k  52 174
## Sample10         n -33  39
## Sample11         a -12  35
## Sample12         c -21  17
## Sample13         k -38 -46
## Sample14         x -57 -28
## Sample15         i -52 -43</code></pre>
</div>
<div id="make-a-phylogenetic-tree-from-the-woodhouse-sequences" class="section level2">
<h2>make a phylogenetic tree from the woodhouse sequences</h2>
<p>To take advantage of the information stored in the DNA sequences we need to generate a phylogenetic tree. Once generated and loaded into the physeq object, phylogeo can use to explore whether there are geogrphic trends in the subclades. The members of the tree must be the same names as the names of the rows (OTUs) in the OTU table.</p>
<pre class="r"><code>#muscle is an APE wrapper for the muscle program
musc &lt;- muscle(woodmouse)
tree &lt;- nj(musc)
plot(tree)</code></pre>
<p><img src="datamodel_files/figure-html/unnamed-chunk-5-1.png" title="" alt="" width="100%" /></p>
</div>
<div id="create-the-phyloseq-object" class="section level2">
<h2>create the phyloseq object</h2>
<p>All of the pieces now must be put together into the phyloseq pbject</p>
<pre class="r"><code>phy &lt;- phyloseq(sample_data = sample_data(sampledata),
                otu_table   = otu_table(otumat, taxa_are_rows=TRUE),
                tax_table   = tax_table(taxmat),
                phy_tree    = phy_tree(tree))
phy</code></pre>
<pre><code>## phyloseq-class experiment-level object
## otu_table()   OTU Table:         [ 15 taxa and 15 samples ]
## sample_data() Sample Data:       [ 15 samples by 9 sample variables ]
## tax_table()   Taxonomy Table:    [ 15 taxa by 7 taxonomic ranks ]
## phy_tree()    Phylogenetic Tree: [ 15 tips and 13 internal nodes ]</code></pre>
</div>
<div id="phylogeo-can-now-be-used-to-map-this-phyloseq-object." class="section level2">
<h2>Phylogeo can now be used to map this phyloseq object.</h2>
<pre class="r"><code># map the locations
# random lat/lons yield samples all over
map_phyloseq(phy)</code></pre>
<p><img src="datamodel_files/figure-html/unnamed-chunk-7-1.png" title="" alt="" width="100%" /></p>
<pre class="r"><code># map the locations
# color by sample data category; size by abundance
map_phyloseq(phy, color=&quot;Category1&quot;, size=&quot;Abundance&quot;)</code></pre>
<p><img src="datamodel_files/figure-html/unnamed-chunk-7-2.png" title="" alt="" width="100%" /></p>
<pre class="r"><code># map the relationship between samples
map_network(phy, color=&quot;Category2&quot;, lines=TRUE)</code></pre>
<p><img src="datamodel_files/figure-html/unnamed-chunk-7-3.png" title="" alt="" width="100%" /></p>
<pre class="r"><code># map the relationship between samples
map_network(phy, color=&quot;Category2&quot;, lines=TRUE)</code></pre>
<p><img src="datamodel_files/figure-html/unnamed-chunk-7-4.png" title="" alt="" width="100%" /></p>
<pre class="r"><code># only keep highly related samples
# tip: if there are unrelated samples that drop out of the network analysis try  base_data = TRUE
map_network(phy, color=&quot;Category2&quot;,lines=TRUE, maxdist = 0.3)</code></pre>
<p><img src="datamodel_files/figure-html/unnamed-chunk-7-5.png" title="" alt="" width="100%" /></p>
<pre class="r"><code># map the tree
# with this random data, perhaps it is not surprisign to wind all of the woodmouse sequences in each of the smapel locations.
map_tree(phy, color=&quot;Category1&quot;)</code></pre>
<p><img src="datamodel_files/figure-html/unnamed-chunk-7-6.png" title="" alt="" width="100%" /></p>
<pre class="r"><code># map clusters
# cluster analyses can let you look at the possilbity of non random distributions of subclades.
# does, say,  a certain subclade only occur in North America?
map_clusters(phy)</code></pre>
<p><img src="datamodel_files/figure-html/unnamed-chunk-8-1.png" title="" alt="" width="100%" /></p>
<pre class="r"><code># plot distance
# pairwise sample distances ecological distance vs. geographic distance
plot_distance(phy)</code></pre>
<p><img src="datamodel_files/figure-html/unnamed-chunk-9-1.png" title="" alt="" width="100%" /></p>
</div>

<footer>
<p><small>The "leaflet" R package is copyright &copy; 2014-2015 RStudio, Inc.<br/>
The Leaflet JavaScript library is &copy; 2010–2014 <a href="http://agafonkin.com/en">Vladimir Agafonkin</a>, 2010–2011 <a href="http://cloudmade.com">CloudMade</a>.<br/>
Maps &copy; <a href="http://openstreetmap.org/copyright">OpenStreetMap</a> contributors unless otherwise noted.
</small></p>
</footer>

  </div>
</div>

<a href="https://github.com/zachcp/phylogeo"><img style="position: absolute; top: 0; right: 0; border: 0;" src="https://camo.githubusercontent.com/38ef81f8aca64bb9a64448d0d70f1308ef5341ab/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6769746875622f726962626f6e732f666f726b6d655f72696768745f6461726b626c75655f3132313632312e706e67" alt="Fork me on GitHub" data-canonical-src="https://s3.amazonaws.com/github/ribbons/forkme_right_darkblue_121621.png"></a>

</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
